extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeStreamExtractor.java

```diff
@@ -830,12 +909,62 @@ private String getEmbeddedInfoStsAndStorePlayerJsUrl() {
                 }
             }
 
-            // Get embed sts
-            return Parser.matchGroup1("\"sts\"\\s*:\\s*(\\d+)", embedPageContent);
-        } catch (final Exception i) {
-            // if it fails we simply reply with no sts as then it does not seem to be necessary
-            return "";
+            if (playerJsUrl != null) {
+                if (playerJsUrl.startsWith("//")) {
+                    playerJsUrl = HTTPS + playerJsUrl;
+                } else if (playerJsUrl.startsWith("/")) {
+                    // Sometimes https://www.youtube.com part has to be added manually
+                    playerJsUrl = HTTPS + "//www.youtube.com" + playerJsUrl;
+                }
+                playerCode = NewPipe.getDownloader().get(playerJsUrl, getExtractorLocalization())
+                        .responseBody();
+            } else {
+                throw new ExtractionException("Could not extract JS player URL");
+            }
+        } catch (final Exception e) {
+            throw new ParsingException("Could not store JavaScript player", e);
+        }
+    }
+
+    private boolean hasOtfStreams() {
+        if (streamingData != null) {
+            boolean hasOtfStreamsValue = false;
+            if (streamingData.has("adaptiveFormats")) {
+                final JsonArray adaptiveFormats = streamingData.getArray("adaptiveFormats");
+                for (final Object adaptiveFormat : adaptiveFormats) {
+                    final JsonObject jsonAdaptiveFormat = (JsonObject) adaptiveFormat;
+                    if (jsonAdaptiveFormat.has("type")) {
+                        final String streamTypeFormat = jsonAdaptiveFormat.getString("type",
+                                EMPTY_STRING);
+                        if (streamTypeFormat.equalsIgnoreCase("FORMAT_STREAM_TYPE_OTF")) {
+                            hasOtfStreamsValue = true;
+                            break;
+                        }
+                    }
+                }
+            }
+            return hasOtfStreamsValue;
+        }
+        return false;
+    }
+
+    private boolean isCipherProtectedContent() {
```

Comment:
Are you sure it is enough to check the first one? Are you sure that either none or all formats are cipher protected?

Reply:
From what I know (with my recent tests), if a stream has a `signatureCipher`, the others are also protected, excepted the DASH manifest (if there is one).

Why Google will protect only some URLs? If they want to make the extraction of a video harder, they will protect all URLs of a stream, I think. But I am not sure.

Let me know what you think and if I need to add a loop in all `adaptiveFormats` and `formats` or not.

Reply:
Yes, I'd prefer you looped through formats. I remember a while ago that some formats were ciphered and some weren't.